contributors/design-proposals/auditing.md

```diff
@@ -0,0 +1,340 @@
+  + has no knowledge of authn, authz, admission
+  + has no access to the storage level for differential output
+  + has to terminate SSL and complicates client certificates based auth
+
+In the following, the second approach is described without a proxy.  At which point there are a few possible places, inside the apiserver, where auditing could happen, namely:
+1. as one of the REST handlers (as in [#27087](https://github.com/kubernetes/kubernetes/pull/27087)),
+2. as an admission controller.
+
+The former approach (currently implemented) was picked over the other one, due to the need to be able to get information about both the user submitting the request and the impersonated user (and group), which is being overridden inside the [impersonation filter](https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apiserver/pkg/endpoints/filters/impersonation.go).  Additionally admission controller does not have access to the response and runs after authorization which will prevent logging failed authorization.  All of that resulted in continuing the solution started in [#27087](https://github.com/kubernetes/kubernetes/pull/27087), which implements auditing as one of the REST handlers
+after authentication, but before impersonation and authorization.
+
+## Proposed Design
+
+The main concepts are those of
+
+- an audit *event*,
+- an audit *level*,
+- an audit *filters*,
+- an audit *output backend*.
+
+An audit event holds all the data necessary for an *output backend* to produce an audit log entry. The *event* is independent of the *output backend*.
+
+The audit event struct is passed through the apiserver layers as an `*audit.Event` pointer inside the apiserver's `Context` object. It is `nil` when auditing is disabled.
+
+If auditing is enabled, the http handler will attach an `audit.Event` to the context:
+
+```go
+func WithAuditEvent(parent Context, e *audit.Event) Context
+func AuditEventFrom(ctx Context) (*audit.Event, bool)
+```
+
+Depending on the audit level (see [below](#levels)), different layers of the apiserver (e.g. http handler, storage) will fill the `audit.Event` struct. Certain fields might stay empty or `nil` if given level does not require that field. E.g. in the case when only http headers are supposed to be audit logged, no `OldObject` or `NewObject` is to be retrieved on the storage layer.
+
+### Levels
+
+Proposed audit levels are:
+
+- `None` - don't audit the request.
+- `Metadata` - reflects the current level of auditing, iow. provides following information about each request: timestamp, source IP, HTTP method, user info (including group as user and as group), namespace, URI and response code.
+- `RequestBody` - additionally provides the unstructured request body.
+- `ResponseBody` - additionally provides the unstructured response body. Equivalent to `RequestBody` for streaming requests.
+- `StorageObject` - provides the object before and after modification or the difference of the changes (to be decided).
```

Comment:
agree on before/after instead of a computed patch, though if storage is in protobuf, this still means two extra JSON encodes of every object per event